src/katgpucbf/xbgpu/engine.py

```diff
@@ -117,6 +121,426 @@ def reset(self, timestamp: int = 0) -> None:
         self.batches = 0
 
 
+class Pipeline:
+    """Base Pipeline class to build on."""
+
+    # NOTE: The n_rx_items and n_tx_items each wrap a GPU buffer. Setting
+    # these values too high results in too much GPU memory being consumed.
+    # There needs to be enough of them that the different processing
+    # functions do not get starved waiting for items. The low single digits
+    # is suitable. n_free_chunks wraps buffer in system RAM. This can be
+    # set quite high as there is much more system RAM than GPU RAM. It
+    # should be higher than max_active_chunks.
+    # These values are not configurable as they have been acceptable for
+    # most tests cases up until now. If the pipeline starts bottlenecking,
+    # then maybe look at increasing these values.
+    # TODO: Declare these constants elsewhere
+    n_rx_items = 3  # Too high means too much GPU memory gets allocated
+    n_tx_items = 2  # TODO: Will likely need to be == n_rx_items for BPipeline
+
+    def __init__(self, output: Output, engine: "XBEngine", context: AbstractContext, init_tx_enabled: bool) -> None:
+        self.engine = engine
+        self.output = output
```

Comment:
Indeed, apologies for that. Do you think it's worth making this base Pipeline default to accepting `list[Output]`
* Then, in XPipeline, only take the first element of that list?
* This might add unnecessary complexity/force me to think about number of/relationship between `rx_item_queue`s vs `tx_item_queue`s.

I'm separately trying to figure out the neatest way of letting the base Pipeline (and `add_rx_item`, `shutdown`) know about the `rx_item_queue` - without adding a/nother `type: ignore`.

Reply:
> Do you think it's worth making this base Pipeline default to accepting list[Output]

It depends. Does the base class (or anything that interacts with pipelines without knowing what sort they are) need to access the (list of) outputs? If so, then your approach is what I'd go with. If not, you could just eliminate the output member from the base class entirely.

> I'm separately trying to figure out the neatest way of letting the base Pipeline (and add_rx_item, shutdown) know about the rx_item_queue

I'm not sure I follow. The `_rx_item_queue` is assigned in `Pipeline.__init__` - does that not achieve it?

Reply:
Agreed on all accounts, but I think it's best if we chat through this one for a few minutes. I'll take you up on your offer of a consult tomorrow morning.

Basically,
* The creation of the item queues requires the `output.name`.
* It doesn't really make sense (to me) to remove the `output` arg (the origin of this comment) and maintain creation of item queues in `Pipeline.__init__`
    * It would require passing through just the `output.name`
    * Which isn't quite desired given idea of one BPipeline to many BOutputs.